### ISSUE TITLE
cloneNode should return sub type, not Node

### ISSUE THREAD

**Issue by dlee** [microsoft/TypeScript#283]
When I do `someElement.cloneNode()`, I want the returned element to be the same type as `someElement`, but right now it's returning type `Node`.

**Comment by Arnavion** [comment]
@DanielRosenwasser @RyanCavanaugh 

Now that TS has `this` for return types, can this be done?

```typescript
interface Node {
    cloneNode(deep?: boolean): this;
}
```

It'd technically be a lie though, since it returns a new object, not itself. Not sure if that matters.

Otherwise would it be okay to add an overload of `cloneNode` to every `Node` subtype in dom.generated.d.ts ?

**Comment by RyanCavanaugh** [comment]
I think `: this` should be correct here?

**Comment by DanielRosenwasser** [comment]
Wouldn't `cloneNode` be inherited from `Node` itself?

**Comment by Arnavion** [comment]
@RyanCavanaugh 

The thing is `: this` is only really suitable for functions that really do return `this`, not a clone.

```typescript
interface Cloneable {
	clone(): this;
}

class Foo implements Cloneable {
	clone(): this {
		// Error. Can work around by asserting to `any`,
		// but still semantically wrong if not overridden by every derived class.
		return new Foo();
	}
}
```

It doesn't matter as much for `Node` and the in-built HTML elements since they have opaque implementations of `cloneNode` and we can assume the browser implements them correctly. But maybe it'll matter somewhere else (custom elements?) and in general is not correct.

**Comment by xogeny** [comment]
I had a similar issue with.  As I understand it, the issue here is that you cannot return anything except `this` because (perhaps among other things), someone could come along and do this:

```
class Bar extends Foo {
}
```

The problem with this is that then you have this problem:

```
  let foo = new Foo();
  let cf: Foo = foo.clone(); // As you would expect
  let bar = new Bar();
  let bf: Bar = bar.clone(); // Ack! bf is really a Foo, but the compiler thinks it is a Bar
```

However, I have a case where I had a fixed set of classes where I needed to do this clone operation.  If you are willing to be disciplined and ensure that each subclass implements `clone`, I **think** you can safely override the compilers (correct) error message safely with a simple (if not slightly misleading) cast, i.e.,

```
class Foo implements Cloneable {
    clone(): this {
       return new Foo() as this; // Just hope nobody extends Foo without also adding clone.
    }
}
```

As I said, I think this is safe (seemed to work for me).  There are semantic issues with "assignability", I didn't really dig into this to see if there are corner cases where this would break anything.  But as far as I can tell, as long as you follow this discipline (always provide a `clone` method), you don't seem to be breaking any of the semantics.

Comments?

**Comment by mindplay-dk** [comment]
> `this` is only really suitable for functions that really do return `this`, not a clone

Because `this` refers to the actual value of `this`, in the same sense that `true`, `false` or `"foo"` refer to actual values and not just types, right?

> I think you can safely override the compilers (correct) error message safely with a simple (if not slightly misleading) cast

I don't think it would be appropriate to rely on a hack like this in the standard libraries?

Looks like what we need is something like `self` - a pseudo-type that *statically* resolves to the parent type, so that type-hinting for e.g. `cloneNode()` becomes possible:

```
interface Element {
    cloneNode(deep?: boolean): self;
}
```

Or with the `Cloneable` example:

```
interface Cloneable {
    clone(): self;
}

class Foo implements Cloneable {
    clone() {
       return new Foo();
    }
}
```

This ought to be safe, even if another class inherits that method:

```
class Bar extends Foo {
    // ERROR: inherited implementation of 'clone' returns 'Foo', expected 'Bar'
}
```

In other words, `self` is like a "type argument" of sorts, that always passes the current type.

Would that work??

**Comment by myfonj** [comment]
(**JSDoc**) Workaround  / for future reference: use **manual [cast](https://github.com/Microsoft/TypeScript/wiki/JSDoc-support-in-JavaScript#casts)** `/** @type {name} */ (ref)`.

```JavaScript
// cloneNode produces Node, @see https://github.com/Microsoft/TypeScript/issues/283
var a = document.cloneNode(true);
a.location; // Property 'location' does not exist on type 'Node'.
// workaround:
var b = /** @type {Document} */ (document.cloneNode(true));
b.location; // OK
```

----
Learned from #25028.

**Comment by callumlocke** [comment]
5 years old :) Is this ever going to be fixed?

**Comment by IllusionMH** [comment]
Do I understand correctly that solution which involve `this` like below ([Playgroung link](https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgLIE8ByB7AJigbwFgAoZc5EPCABSmwAcAuZAZzClAHMBuUi5AgA22EBBz4APABVkEAB6QQuVmizVkAXjUSIAPgAUYABbBWLaQBpk+CAwD8LAEbZsQiHBABKC3xIBfUlJQSFhEFAwAUXcAWwhwOUV4lR0NYjIKCFj4sDpGFhAAVxinaD9AkmDwaHgkNQBBEARjbChEpRSo7IT0gWMoCBgWdk4QXlIKqtDaiPQASRiuduTVLog4nv4KVigEYY5uPz6IYC5jMEdKYtKoI4oAd2BcE0uikrKJoJJ8YTgBwVE7Eo1BYGF0fh+Qj+KAQgLAcm6YFB6Gi6xyEIgv3+sJAQM8zVayMaBNupEh0IBuPhwEWyIW40qJBxQOEogguF0WmB+AAdKyxLoDF4-Mz4fzkqiNvDtFk0eA+SIBdQhSK4YJFeziS02tp8dqFWzBcLSKL1WzcPSuTSuAalfgVV9TXLcvQGNIIOwCtdoFzxbgta0ebKpXkGKqqdzaK73Z62Acxr6Nf6mvqqPhQ+Ggf1BjGkXHRkttH6A1AedmYH4TWrjCczmBc173m0AD7IQrKQagdmJ830su186Z+GPZ7GBtXJvIVvt-AwLu4HtiC2LHkjkxDti7ccjbiL9l9nYIPxAA))
```ts
cloneNode<T extends Node = Node>(this: T, deep?: boolean): T;
```
 won't be considered and expected fix is to add specific `cloneNode` definitions to all decedents of `Node` interface?

And those who create custom element will have to define their own types for this method?

**Comment by fregante** [comment]
Is it not possible to extract the class type from `this`? Example:

```ts
interface Node {
    cloneNode(deep?: boolean): Instance<Constructor<this>>; // Not real
}
```

**Comment by IllusionMH** [comment]
@fregante see https://github.com/microsoft/TSJS-lib-generator/pull/302#discussion_r143547060 . It is working solution (see comment above for sorter version), but "That PR would make compile time and memory usage worse for everybody, so we can't take it." (even if they don't need `cloneNode`). So if you need it - add to your repo, but be aware of drawbacks.

PR that adds overrides with concrete types is in place and waiting for review.

**Comment by fregante** [comment]
> So if you need it - add to your repo, but be aware of drawbacks.

Right, that's what I was looking to solve.

> PR that adds overrides with concrete types is in place and waiting for review.

Oh, I see that now.	Easy solution but does it work correctly on custom elements?

**Comment by IllusionMH** [comment]
> does it work correctly on custom elements

No, custom elements declaration would have return type of element that they extend (which will be close enough for some cases) but need own override to have exact type :(

I'm not sure that it is possible to do so without making `cloneNode` generic, but it can be done on project level if needed.

**Comment by piotr-cz** [comment]
As in @myfonj example, the trick is to wrap expression in brackets `()`:

```js
/** @type {HTMLTableRowElement} */
const $templateRow = document.createElement('tr')

/** @type {HTMLTableRowElement} */
const $rowClone1 = $templateRow.cloneNode(true)
// Error: `Type 'Node' is missing the following properties from type 'HTMLTableRowElement': ...`

/** @type {HTMLTableRowElement} */
const $rowClone2 = ($templateRow.cloneNode(true))
// No errors
```

**Comment by sandersn** [comment]
Unfortunately, we had to revert microsoft/TSJS-lib-generator#811 because it makes type parameters that extend HTMLElement invariant where they were previously covariant. See https://github.com/microsoft/TSJS-lib-generator/pull/842

`@types/tablesorter` is a good test case for future attempts; clone microsoft/DefinitelyTyped and then run `tsc` in types/tablesorter.

**Comment by clshortfuse** [comment]
This works:

````ts
interface Node extends EventTarget {
    cloneNode(deep?: boolean): this;
}
````

Not sure if `this` wasn't supported before, but seems fine now. `DocumentFragment` and `Element` and `Node` all clone their respective types. Even mixed typed (eg: (`DocumentFragment|Element`).

The only problem I see is `ChildNode.cloneNode()` should return `Node`, though, there's a probably a way to coerce that in the `ChildNode` interface.

**Comment by clshortfuse** [comment]
There's also `typeof this` which sounds the same, but isn't. (See https://github.com/microsoft/TypeScript/issues/51265)

There's not much information about the performance drain, so I can't speak to if `typeof this` would be any different.

**Comment by callionica** [comment]
This continues to be a problem. Could someone reassess fixing this? If there are still problems with more complete solutions, please add a specific version of cloneNode for HTMLElement. If cloneNode returned HTMLElement from an HTMLElement (even if no changes for more specific types), it would still fix a bunch of code and be a huge usability improvement.

**Comment by miss-programgamer** [comment]
Seems to me like typescript needs a generalized way of expressing the concept of "the type of the current class *or* whichever one of its subtypes, depending on context". Having `this` as a special return type is a start, but it only handles a very specific use case. Maybe something like `typeof class` could be used to express this? It otherwise means nonsense (unlike `typeof this`) and is short enough to be ergonomic to type.

Theoretical usage example:

```ts
interface Node {
    cloneNode(): typeof class;
}
```

Or, for a more complex use case I've encountered myself:

```ts
abstract class GameNode {
     emplace(ComponentConstructor: new(node: typeof class) => Component): Component {
        return new ComponentConstructor(this);
    }
}
```

Important note about the snippet above is that `this` would need to be implicitly convertible to `typeof class` and I'm unsure if that could cause problems in other ways. If so, I could live with having to do `this as typeof class` when appropriate.

**Comment by denis-migdal** [comment]
Wouldn't it be possible to redefine `cloneNode` at least for `DocumentFragment` and `Element`/`HTMLElement` ?

It won't be perfect, but would be better than nothing, without having to make Node generic or to redefine every classes inheriting from Node.

```ts
declare global {
    interface DocumentFragment {
        cloneNode(deep?: boolean): DocumentFragment;
    }
    interface HTMLElement {
        cloneNode(deep?: boolean): HTMLElement;
    }
}
```

For information, you can also do :
```ts
declare global {
    interface Node {
        cloneNode(deep?: boolean): this;
    }
}
```
